crates/rust-analyzer/src/bin/main.rs

```diff
@@ -104,42 +108,20 @@ fn setup_logging(log_file: Option<&Path>, no_buffering: bool) -> Result<()> {
         None => None,
     };
     let filter = env::var("RA_LOG").ok();
-    logger::Logger::new(log_file, no_buffering, filter.as_deref()).install();
-
-    tracing_setup::setup_tracing()?;
+    let guard = logger::Logger::new(log_file, filter.as_deref()).install();
 
     profile::init();
 
-    Ok(())
-}
-
-mod tracing_setup {
-    use tracing::subscriber;
-    use tracing_subscriber::layer::SubscriberExt;
-    use tracing_subscriber::EnvFilter;
-    use tracing_subscriber::Registry;
-    use tracing_tree::HierarchicalLayer;
-
-    pub(crate) fn setup_tracing() -> super::Result<()> {
-        let filter = EnvFilter::from_env("CHALK_DEBUG");
```

Comment:
Did we lose this?

Reply:
I suppose this is mostly to get `tracing-subscriber` to work  properly, and now it already works. We can just put some directions in `RA_LOG` to tell chalk module how to output its logs. Maybe I missed something.

Reply:
Can you check if the output is similar? We might e.g. be losing indentation or the other settings. I think we can always enable them, though.

> We can just put some directions in RA_LOG to tell chalk module how to output its logs.

Yeah, but we'll also have to set `CHALK_DEBUG` for `is_chalk_debug` and `crates/hir_ty/src/tests.rs`. I'm not sure how it's best to handle this.

Reply:
I'm gonna to do more test on it.

Reply:
Hi Nicola, after two days of struggling, I can't find the best solution to this problem. Here are some scenarios:

1. If we have to read two environment variables from `RA_LOG` and `CHALK_DEBUG`  to determine how the filter works, what's the relationship between the two? Previously, people might set `CHALK_DEBUG=debug` (actually this makes the whole program to print debug log), it works because there is just only some chalk crates use tracing, so the log written to stderr looks appropriate.
2. If we treat `CHALK_DEBUG` only affect to chalk package by manually add directive like `chalk_solve={CHALK_DEBUG}` to main filter (from `RA_LOG`), it works properly. But how do we write chalk debug log to stderr and ra_log to a file or stderr (if we set `RA_LOG=info CHALK_DEBUG=debug`, the stderr will be filled up with logs from `tracing-tree`, as `tracing-tree` also output other crate's log.)? Perhaps we should implement a custom `Subscriber` or modify some code in `tracing-tree` project to make it possible to write only the logs of specified targets to stderr or elsewhere.
3. Should we copy the `tracing-tree` code to this repo and modify it, or do something else? Or it is acceptable, because it behaves pretty much the same as the original except for the problem I said in 2 (set `RA_LOG=info CHALK_DEBUG=debug`, massive logs in stderr).

I have no idea how to do right now, and I have submitted the code as I said above. Could you give me some suggestions? Thank you a lot.

Reply:
I don't have any experience with `tracing`, so I might be missing something, but what's the problem with 2?

I thought we could do one of these:

 - ask the user to set something like `RA_LOG=error,chalk_solve=debug,etc`, detect that logging is enabled for `chalk_solve` and do wire things up for `is_chalk_debug` and friends
 - detect `CHALK_DEBUG` and add `chalk_solve=debug,etc` to `RA_LOG` before setting up `tracing`

> But how do we write chalk debug log to stderr and ra_log to a file or stderr

Why is this a problem? If you ask for `rust-analyzer` and `chalk` logs, you get logs for both. You can set up a proper filter to make them less verbose.

Reply:
Between the two above, I'd go for the first one, it seems overall simpler and more predictable.

Reply:
I think for `CHALK_DEBUG`, it'd be fine to either make it append `chalk_solve=x` to `RA_LOG` or just do nothing. `is_chalk_debug` can just check whether logging is enabled.

Reply:
> Why is this a problem? If you ask for rust-analyzer and chalk logs, you get logs for both. You can set up a proper filter to make them less verbose.

I mean, `tracing-tree` collects all the events and spans issued by `tracing` and prints them in tree-structured format. If we both set `RA_LOG` and `CHALK_DEBUG`, all logs issued by `tracing` will write to both `stderr` and a log file (if I had set --log-file). Do you think it's a problem? Maybe it's not.

> Or do you mean that if you set a log file, the chalk logs will go there instead of on stderr like before?

Yes, not only chalk logs will go there, but also ra_log will go `stderr`, if the pattern is matched.

Reply:
@flodiebold Agree, that's what I'm doing now.

Reply:
Yeah, that's exactly what I'd expect, all the logs going to the same place.

Reply:
Okay, so is there any issues I should fix?